### ISSUE TITLE
`__builtin_dynamic_object_size` does not report size of function argument VLAs

### ISSUE THREAD

**Issue by kees** [llvm/llvm-project#151240]
For function argument VLAs, GCC correctly reports the expected size. Clang seems unaware:

```
#include <stdio.h>

void __attribute__((noinline))
foo(int n, short array[n])
{
        printf("%zu\n", __builtin_dynamic_object_size(array, 1));
}

int main(void)
{
        short array[10];

        foo(10, array);
        return 0;
}
```

```
$ gcc -Wall -O2 -o fvla-gcc fvla.c && ./fvla-gcc
20
$ clang -Wall -O2 -o fvla-clang fvla.c && ./fvla-clang
18446744073709551615
```

@bwendling

**Comment by efriedma-quic** [comment]
Don't you need `foo(int n, short array[static n])`?  By default, the bound of an array before adjustment doesn't have any semantic effect.

**Comment by kees** [comment]
That shouldn't matter (this is about the runtime size: bos has no idea with or without `static`, but bdos has what it needs to figure it out). And Clang doesn't emit the right answer in that case either. :P

**Comment by efriedma-quic** [comment]
Let's simplify:

```
#include <stdio.h>
void foo(short array[10])
{
        printf("%zu\n", __builtin_object_size(array, 1));
}
```

clang and gcc both agree the size is unknown, because this is semantically exactly the same as `void foo(short* array)`.

I don't see how bdos or VLAs change that analysis.

**Comment by efriedma-quic** [comment]
You're saying that bdos has some special semantics that say we should pretend we know the size of the object pointed to by pointers declared in a certain way, even though there's no semantic rule that actually specifies the size, and there's no documentation anywhere that says this is supposed to happen?

I think this breaks our normal rules for new extensions, which is that they have to have some reserved extension marker.  I mean, technically there's a marker when you call `bdos`, but that's not helpful when someone's code breaks because FORTIFY starts using `bdos` internally.

**Comment by kees** [comment]
I feel like we're talking past each other, but this is _for_ use with FORTIFY: it is the primary (only?) user of `bdos`. It is expecting to tie declared object size information to pointers, and this is what we have here: the variable is an array with a specific size. FORTIFY is just waiting for `bdos` to gain this awareness. The whole point of the bounds-safety work is to fix these gaps in object size awareness so that `bdos` and the sanitizers gain access to the size information that has been semantically expressed in the language.

**Comment by efriedma-quic** [comment]
You can't just declare a standard C construct "specifies the size" when the standard says it does not, in fact, specify the size.  Like, we can't say `int ((*x));` points to an object of exactly 4 bytes because there are four parentheses in the declarator.

**Comment by kees** [comment]
But `bdos` is not part of the standard. We can make changes to language extensions, especially to maintain interoperability with GCC.

**Comment by kees** [comment]
> You can't just declare a standard C construct "specifies the size" when the standard says it does not, in fact, specify the size. Like, we can't say `int ((*x));` points to an object of exactly 4 bytes because there are four parentheses in the declarator.

Also, this seems like an unreasonable slippery slope argument. If source code shows an array of size 4, it is perfectly reasonable to expect that it is, in fact, an array of size 4.

**Comment by efriedma-quic** [comment]
bdos is not part of the standard, sure, but the impact is non-local.  It's likely that user code will be implicitly opted-in to using it using FORTIFY, or it'll be used in an inline function in some library outside the user's control.  Which means any code containing array bounds in a function definition must follow your non-standard rule, or it will randomly break.

If someone writes `foo(int n, short array[__counted_by(n)])` or something like that, they've explicitly opted into your semantics using a reserved keyword.  If they haven't, you're basically declaring that gcc is going to ignore whatever the standards committee says in this area.

Which... if we think this extension is important enough, we can ignore the committee, I guess.  But I'd at least like some documentation explaining how exactly we're deviating from the standard.

**Comment by alejandro-colomar** [comment]
> Don't you need `foo(int n, short array[static n])`? By default, the bound of an array before adjustment doesn't have any semantic effect.

This is a defect in the standard.  Quality implementations, including GCC, consider `[n]` to be as good as `[static n]`.  In those implementations, `static` only means non-null.

```c
alx@devuan:~/tmp$ cat n.c 
void f(int a[10]);
void f(int a[20]);
alx@devuan:~/tmp$ gcc -Wall -S n.c 
n.c:2:12: warning: argument 1 of type ‘int[20]’ with mismatched bound [-Warray-parameter=]
    2 | void f(int a[20]);
      |        ~~~~^~~~~
n.c:1:12: note: previously declared as ‘int[10]’
    1 | void f(int a[10]);
      |        ~~~~^~~~~
```

There's work going on to fix that in the standard.
